File: app/src/main/java/pl/pregiel/workwork/fragments/AddWorkTimeFragment.java
package pl.pregiel.workwork.fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.Spinner;

import pl.pregiel.workwork.ControlSetup;
import pl.pregiel.workwork.R;
import pl.pregiel.workwork.Settings;
import pl.pregiel.workwork.Utils;
import pl.pregiel.workwork.data.database.services.WorkService;
import pl.pregiel.workwork.data.database.services.WorkTimeService;
import pl.pregiel.workwork.data.pojo.Work;
import pl.pregiel.workwork.data.pojo.WorkTime;
import pl.pregiel.workwork.exceptions.EmptyFieldException;
import pl.pregiel.workwork.exceptions.ShowToastException;
import pl.pregiel.workwork.utils.ErrorToasts;


public class AddWorkTimeFragment extends FormFragment {
    public static final String TAG = "ADD_WORKTIME";

    private WorkTimeService workTimeService;
    private WorkService workService;

    private Work work;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle arguments = getArguments();
        workService = new WorkService(getContext());
        workTimeService = new WorkTimeService(getContext());

        if (arguments != null && getActivity() != null) {
            work = workService.getById(arguments.getInt("work_id"));
            getActivity().setTitle(String.format(getString(R.string.title_addWorkTime), work.getTitle()));
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_addworktime, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        final EditText dateEditText = view.findViewById(R.id.editText_addWorkTime_date);
        ControlSetup.setupDatePicker(getContext(), dateEditText);

        final EditText timeFromEditText = view.findViewById(R.id.editText_addWorkTime_from);
        final EditText timeToEditText = view.findViewById(R.id.editText_addWorkTime_to);

        if (work.getTimeMode() == 1) {
            ControlSetup.setupTimePickerWithDelay(getContext(), timeFromEditText, work.getTimeToNow());
            ControlSetup.setupTimePicker(getContext(), timeToEditText);
        } else {
            ControlSetup.setupTimePicker(getContext(), timeFromEditText, work.getTimeFrom());
            ControlSetup.setupTimePicker(getContext(), timeToEditText, work.getTimeTo());
        }

        final EditText timeAmountEditText = view.findViewById(R.id.editText_addWorkTime_amount);
        ControlSetup.setupTimeAmountPicker(getContext(), timeAmountEditText, work.getTimeAmount());

        final RadioButton timeRangeRadioButton = view.findViewById(R.id.radioButton_addWorkTime_rangeTime);
        final RadioButton timeAmountRadioButton = view.findViewById(R.id.radioButton_addWorkTime_amountTime);
        ControlSetup.setupCustomRadioButtonGroup(timeRangeRadioButton, timeAmountRadioButton);

        final EditText salaryEditText = view.findViewById(R.id.editText_addWorkTime_salary);
        ControlSetup.setupSalaryEditText(salaryEditText, String.valueOf(Settings.DEFAULT_HOURLY_WAGE));

        final RadioButton salaryPerHourRadioButton = view.findViewById(R.id.radioButton_addWorkTime_salaryPerHour);
        final Spinner currencySpinner = view.findViewById(R.id.spinner_addWorkTime_currency);
        ControlSetup.setupCurrencySpinner(getContext(), currencySpinner, work.getCurrency());

        final EditText infoEditText = view.findViewById(R.id.editText_addWorkTime_info);

        final Button addButton = view.findViewById(R.id.button_addWorkTime_add);
        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                try {
                    WorkTime workTime = new WorkTime();
                    workTime.setDay(dateEditText.getText().toString());
                    if (timeRangeRadioButton.isChecked()) {
                        int timeFromInMinutes = Utils.timeStringToMinutes(timeFromEditText.getText().toString());
                        int timeToInMinutes = Utils.timeStringToMinutes(timeToEditText.getText().toString());

                        workTime.setTimeFrom(timeFromInMinutes);
                        workTime.setTimeTo(timeToInMinutes);
                        workTime.setTime(-1);

                        if (timeFromInMinutes < timeToInMinutes) {
                            workTime.setTime(timeToInMinutes - timeFromInMinutes);
                        } else {
                            workTime.setTime((1440 + timeToInMinutes) - timeFromInMinutes);
                        }
                    } else {
                        int timeAmountInMinutes = Utils.timeStringWithSuffixesToMinutes(timeAmountEditText.getText().toString());

                        workTime.setTimeFrom(-1);
                        workTime.setTimeTo(-1);
                        workTime.setTime(timeAmountInMinutes);
                    }

                    String salaryString = salaryEditText.getText().toString().replace(",", ".");
                    if (salaryString.isEmpty()) {
                        throw new EmptyFieldException(getString(R.string.global_salary));
                    }

                    int salary = Math.round(Float.valueOf(salaryString) * 100);
                    workTime.setSalary(salary);
                    workTime.setSalaryMode(salaryPerHourRadioButton.isChecked() ? 0 : 1);
                    workTime.setInfo(infoEditText.getText().toString());
                    workTime.setWorkId(work.getId());

                    workTimeService.create(workTime);

                    if (work.getCurrency() != currencySpinner.getSelectedItemPosition()) {
                        work.setCurrency(currencySpinner.getSelectedItemPosition());
                        workService.update(work);
                    }
                    if (getContext() != null) {
                        ((FragmentActivity) getContext()).onBackPressed();
                    }
                } catch (ShowToastException e) {
                    e.showToast(getContext());
                } catch (Exception e) {
                    ErrorToasts.showUnknownErrorToast(getContext());
                }
            }
        });

        final Button cancelButton = view.findViewById(R.id.button_addWorkTime_cancel);
        cancelButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (getContext() != null) {
                    ((FragmentActivity) getContext()).onBackPressed();
                }
            }
        });
    }
}

File: app/src/main/java/pl/pregiel/workwork/fragments/AddWorkTimeFragment.kt
package pl.pregiel.workwork.fragments

import android.os.Bundle
import android.support.v4.app.FragmentActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.fragment_addworktime.*
import pl.pregiel.workwork.R
import pl.pregiel.workwork.Settings
import pl.pregiel.workwork.data.database.services.WorkService
import pl.pregiel.workwork.data.database.services.WorkTimeService
import pl.pregiel.workwork.data.pojo.Work
import pl.pregiel.workwork.data.pojo.WorkTime
import pl.pregiel.workwork.exceptions.EmptyFieldException
import pl.pregiel.workwork.exceptions.ShowToastException
import pl.pregiel.workwork.utils.ControlSetup
import pl.pregiel.workwork.utils.ErrorToasts
import pl.pregiel.workwork.utils.Utils
import kotlin.math.roundToInt


class AddWorkTimeFragment : FormFragment(), TaggedFragment {

    override fun fragmentTag() = "ADD_WORKTIME"

    private lateinit var workService: WorkService
    private lateinit var workTimeService: WorkTimeService

    private lateinit var work: Work

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val arguments = arguments
        workService = WorkService(context!!)
        workTimeService = WorkTimeService(context!!)

        work = workService.getById(arguments!!.getInt("work_id"))!!
        activity?.title = String.format(getString(R.string.title_addWorkTime), work.title)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_addworktime, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        ControlSetup.setupDatePicker(context!!, editText_addWorkTime_date)
        ControlSetup.setupTimeAmountPicker(context!!, editText_addWorkTime_amount, work.timeAmount)
        ControlSetup.setupCustomRadioButtonGroup(radioButton_addWorkTime_rangeTime, radioButton_addWorkTime_amountTime)
        ControlSetup.setupSalaryEditText(editText_addWorkTime_salary, Settings.DEFAULT_HOURLY_WAGE.toString())
        ControlSetup.setupCurrencySpinner(context!!, spinner_addWorkTime_currency, work.currency)

        if (work.timeMode == 1) {
            ControlSetup.setupTimePickerWithDelay(context!!, editText_addWorkTime_from, work.timeToNow)
            ControlSetup.setupTimePicker(context!!, editText_addWorkTime_to)
        } else {
            ControlSetup.setupTimePicker(context!!, editText_addWorkTime_from, work.timeFrom)
            ControlSetup.setupTimePicker(context!!, editText_addWorkTime_to, work.timeTo)
        }

        button_addWorkTime_add.setOnClickListener {
            try {
                val workTime = WorkTime()
                workTime.day = editText_addWorkTime_date.text.toString()
                if (radioButton_addWorkTime_rangeTime.isChecked) {
                    val timeFromInMinutes = Utils.timeStringToMinutes(editText_addWorkTime_from.text.toString())
                    val timeToInMinutes = Utils.timeStringToMinutes(editText_addWorkTime_to.text.toString())

                    workTime.timeFrom = timeFromInMinutes
                    workTime.timeTo = timeToInMinutes
                    workTime.time = -1

                    workTime.time = if (timeFromInMinutes < timeToInMinutes)
                        timeToInMinutes - timeFromInMinutes
                    else
                        1440 + timeToInMinutes - timeFromInMinutes
                } else {
                    val timeAmountInMinutes = Utils.timeStringWithSuffixesToMinutes(editText_addWorkTime_amount.text.toString())

                    workTime.timeFrom = -1
                    workTime.timeTo = -1
                    workTime.time = timeAmountInMinutes
                }

                val salaryString = editText_addWorkTime_salary.text.toString().replace(",", ".")
                if (salaryString.isEmpty()) {
                    throw EmptyFieldException(getString(R.string.global_salary))
                }

                workTime.salary = (salaryString.toDouble() * 100).roundToInt()
                workTime.salaryMode = if (radioButton_addWorkTime_salaryPerHour.isChecked) 0 else 1
                workTime.info = editText_addWorkTime_info.text.toString()
                workTime.workId = work.id

                workTimeService.create(workTime)

                if (work.currency != spinner_addWorkTime_currency.selectedItemPosition) {
                    work.currency = spinner_addWorkTime_currency.selectedItemPosition
                    workService.update(work)
                }

                if (context != null) {
                    (context as FragmentActivity).onBackPressed()
                }
            } catch (e: ShowToastException) {
                e.showToast(context)
            } catch (e: Exception) {
                ErrorToasts.showUnknownErrorToast(context)
            }
        }

        button_addWorkTime_cancel.setOnClickListener {
            if (context != null) {
                (context as FragmentActivity).onBackPressed()
            }
        }
    }
}
